server/routes/api/documents.test.ts

```diff
@@ -15,9 +15,15 @@ import {
   buildDocument,
   buildViewer,
 } from "@server/test/factories";
-import { flushdb, seed, getTestServer } from "@server/test/support";
+import {
+  flushdb,
+  seed,
+  getTestServer,
+  disconnectdb,
+} from "@server/test/support";
 
 const server = getTestServer();
+afterAll(() => disconnectdb().then(() => server.close()));
```

Comment:
This plugged up the leaks detected using `--detect-leaks` flag as shown [here](https://github.com/outline/outline/issues/3939#issuecomment-1209012261).

Reply:
I recently (as part of this issue) changed this to use a single centralized test server – the thinking was that it would be fast than spinning up an instance/dbconnection in each file. I'm actually pretty surprised this works though and the rest of the suite doesn't fail after the server is closed? 🤔

Reply:
> I'm actually pretty surprised this works though and the rest of the suite doesn't fail after the server is closed?

Looks like [this](https://github.com/outline/outline/blob/e5c5e8907a52e82356912a5403a6321f2682edd1/server/test/support.ts#L119) is unreachable. Not sure why :thinking:

Reply:
Oh, I'll admit I didn't explicitly check that. Must be a limitation of jest module scope, in that case how about we change `getTestServer` to the following:

```
export function getTestServer() {
  const app = webService();
  const testServer = new TestServer(app.callback());

  testServer.cleanup = async () => {
    await sequelize.close();
    await testServer.close();
  };

  return testServer;
}
```

Then we just need to add the following in individual test files

```
afterAll(server.cleanup);
```

Reply:
You're right though, this cleanup makes a huge difference 🙏

Reply:
> Not sure why, likely due to how jest handles running multiple test files :thinking:

I think we need [this](https://jestjs.io/docs/configuration#setupfilesafterenv-array).

Reply:
Maybe, but not all the server tests need a TestServer instance – which means when you're iterating on eg a command test it's running the server for no reason. Makes me prefer the cleanup approach

Reply:
> Must be a limitation of jest module scope.

Yeah, looks like there isn't a direct way to facilitate state sharing across different test files in `jest`.

Reply:
```
export function getTestServer() {
  const app = webService();
  const testServer = new TestServer(app.callback());

  testServer.cleanup = async () => {
    await sequelize.close();
    await testServer.close();
  };

  return testServer;
}
```
Err...we're aren't importing `getTestServer` in all of the `server` tests :) :point_down: 

> but not all the server tests need a TestServer instance

Reply:
I was replying to your comment about potentially using jest `setupFilesAfterEnv`

Reply:
Ahh..I thought that facilitated state sharing across different test files, but it doesn't seem to(so, I deleted the comment :laughing: )

Reply:
One day we'll enjoy modern luxuries like thread replies on GitHub

Reply:
I'm thinking maybe it's better to do the cleanup in each of the individual test files itself. Although, there would be some repetition, but it's fine to repeat then have a potentially brittle abstraction? Let me know what do you think.

Reply:
The more we can centralize setup and teardown the less future similar mistakes, I'm okay centralizing – particularly as moving into the suites would also require importing sequelize

Reply:
How about :point_down: 

```
export function getTestServer() {
  const app = webService();
  const server = new TestServer(app.callback());

  server.teardown = () => {
    await sequelize.close();
    await server.close();
  };

  return server;
}

export function getTestdb() {
  const db = sequelize;

  db.flush = () => { /* flush! */ };
  db.seed = () => { /* seed! */ };
  db.disconnect () => { /* disconnect! */ };
};
```

And,

```
// Usage(in server/routes/api/)

import { getTestServer, getTestdb } from "@server/test/utils"

const db = getTestdb();
const server = getTestServer();

afterAll(server.disconnect);

beforeEach(db.flush);
beforeEach(db.seed);

//////////////////////////////////////////////////////////////////////////

// Usage(in server/{commands, models...})
import { getTestdb } from "@server/test/utils";


const db = getTestdb();

afterAll(db.disconnect);

beforeEach(db.flush);
beforeEach(db.seed);
```

Reply:
Okay, but `getTestdb` -> `getTestDatabase`

And don't include `seed`, I'm trying to gradually get rid of it in favor of individual factories.

Reply:
> And don't include seed, I'm trying to gradually get rid of it in favor of individual factories.

So, we keep `seed` as it in current scope? Also, are you exploring an equivalent of [`factory_bot`](https://github.com/thoughtbot/factory_bot)? :)

Reply:
Okay – include seed if you're prepared to make the changes 😆